### PR TITLE
New debug and feature flag infrastructure

### PR DESCRIPTION
**Pull request by chadhietala**
This introduces new [code stripping infrastructure](https://github.com/chadhietala/babel-plugin-debug-macros) to lead the way for Project Svelte.

### CODE REVIEW

File: packages/container/lib/container.js [hunk 1]
```diff
@@ -126,7 +130,7 @@ Container.prototype = {
   lookupFactory(fullName, options) {
     assert('fullName must be a proper full name', this.registry.validateFullName(fullName));
 
-    deprecate('Using "_lookupFactory" is deprecated. Please use container.factoryFor instead.', !isFeatureEnabled('ember-factory-for'), { id: 'container-lookupFactory', until: '2.13.0', url: 'http://emberjs.com/deprecations/v2.x/#toc_migrating-from-_lookupfactory-to-factoryfor' });
+    deprecate('Using "_lookupFactory" is deprecated. Please use container.factoryFor instead.', false, { id: 'container-lookupFactory', until: '2.13.0', url: 'http://emberjs.com/deprecations/v2.x/#toc_migrating-from-_lookupfactory-to-factoryfor' });
```
Comment (chadhietala) [inline packages/container/lib/container.js]: Fix this

File: package.json [hunk 2]
```diff
@@ -40,6 +40,7 @@
     "@glimmer/reference": "^0.22.0",
     "@glimmer/runtime": "^0.22.0",
     "@glimmer/util": "^0.22.0",
+    "babel-plugin-minify-dead-code-elimination": "^0.1.4",
```
Comment (chadhietala) [inline package.json]: Fix this

File: broccoli/packages.js [hunk 3]
```diff
@@ -9,9 +9,11 @@ const funnelLib = require('./funnel-lib');
 const { VERSION } = require('./version');
 const WriteFile = require('broccoli-file-creator');
 const StringReplace = require('broccoli-string-replace');
-const { RELEASE, DEBUG } = require('./features');
+const { RELEASE, DEBUG, toConst } = require('./features');
```
Comment (krisselden) [inline broccoli/packages.js]: is RELEASE not just !DEBUG?

Comment (krisselden) [inline broccoli/packages.js]: nevermind, I see this isn't a bool

File: packages/container/lib/container.js (hunk 1 above)
Comment (chadhietala) [inline packages/container/lib/container.js]: Done.

**Comment by rwjblue** [comment]
Can you share a tarball of a prod build so we can review?

File: broccoli/packages.js [hunk 4]
```diff
@@ -233,7 +235,16 @@ module.exports.emberLicense = function _emberLicense() {
...
 
 module.exports.emberFeaturesES = function _emberFeaturesES(production = false) {
-  let content = 'export default ' + JSON.stringify(production ? RELEASE : DEBUG) + ';\n';
+  let FEATURES = production ? RELEASE : DEBUG;
+  let content = stripIndent`
+    const FEATURES = ${JSON.stringify(FEATURES)};
+    export default FEATURES;
+
+    ${Object.keys(FEATURES).map((feature) => {
+      return `export const ${feature.replace(/-/g, '_').toUpperCase()} = FEATURES["${feature}"];`
```
Comment (rwjblue) [inline broccoli/packages.js]: Isn't this what toConst was for?

File: broccoli/to-es5.js [hunk 5]
```diff
@@ -37,14 +48,11 @@ module.exports = function toES5(tree, _options) {
...
-    options.plugins.push([
-      'filter-imports', {
-        'ember-debug/deprecate': ['deprecate'],
-        'ember-debug': ['assert', 'debug', 'deprecate', 'info', 'runInDebug', 'warn', 'debugSeal', 'debugFreeze']
-      }
-    ]);
+    options.plugins.push(['minify-dead-code-elimination', { 'optimizeRawSize': true }]);
   }
 
+  // options.plugins.push(['transform-es2015-block-scoping',  { 'throwIfClosureRequired': true }]);
```
Comment (rwjblue) [inline broccoli/to-es5.js]: What is this comment about?

File: broccoli/to-es5.js [hunk 6]
```diff
@@ -3,20 +3,31 @@
...
 const Babel = require('broccoli-babel-transpiler');
 const injectBabelHelpers = require('./transforms/inject-babel-helpers');
-const { RELEASE, DEBUG } = require('./features');
+const { RELEASE, DEBUG, toConst } = require('./features');
 
 module.exports = function toES5(tree, _options) {
   let options = Object.assign({
     environment: 'developement'
   }, _options);
+  options.persist = false;
```
Comment (rwjblue) [inline broccoli/to-es5.js]: Do we really want to disable cross restart caching?

File: packages/container/tests/container_test.js [hunk 7]
```diff
@@ -570,65 +570,6 @@ QUnit.test('A deprecated `container` property is appended to every object instan
   }, 'Using the injected `container` is deprecated. Please use the `getOwner` helper instead to access the owner of this object.');
 });
 
-// This is testing that container was passed as an option
-QUnit.test('A deprecated `container` property is appended to every object instantiated from a non-extendable factory, and a fake container is available during instantiation.', function() {
```
Comment (rwjblue) [inline packages/container/tests/container_test.js]: This test should not be removed. The container deprecation should still be tested.

File: packages/ember-views/lib/system/utils.js [hunk 8]
```diff
@@ -187,5 +187,6 @@ export const elMatches = typeof Element !== 'undefined' &&
    Element.prototype.webkitMatchesSelector);
 
 export function matches(el, selector) {
+  debugger;
```
Comment (rwjblue) [inline packages/ember-views/lib/system/utils.js]: Left over from debugging?

**Review by rwjblue — COMMENTED** [review]
Can you explain how features get enabled at runtime in both debug and prod builds in the new system (without rebuilding)?

File: packages/container/tests/container_test.js (hunk 7 above)
Comment (rwjblue) [inline packages/container/tests/container_test.js]: Chad pointed out that this test was already basically skipped just below.

**Comment by mmun** [comment]
This is really cool. I love that we can lean on dead code elimination for more complicated feature flagging combinations! :)

**Comment by rwjblue** [comment]
It does not seem possible to enable features at runtime with this at the moment, with this branch (but it does work on canary builds).

[Demo JSBIN against locahost:4200](http://jsbin.com/gubesom/edit?html,js,output)
[Demo JSBIN against canary](http://jsbin.com/zawiris/edit?html,js,output)

**Pull request merged**